### PR TITLE
vmm: cpu: Add pending removed vcpu check to avoid resize vcpu hang

### LINKED ISSUE #5419: VM hangs when stressing CPU hotplug
**Describe the bug**
When repeatedly resizing a VM with different numbers of virtual CPUs, from 1 to 4, in a loop, the VM will hang after a few (<10) iterations. 

```bash
$ while true; do echo "Stressing from 1 to 4"; for i in $(seq 1 4); do curl  -X PUT -d"{\"desired_vcpus\": $i}" -H 'Accept: application/json' --unix-socket /tmp/cloud-hypervisor.sock  http://localhost/api/v1/vm.resize; done; done
Stressing from 1 to 4
Stressing from 1 to 4
Stressing from 1 to 4
Stressing from 1 to 4
Stressing from 1 to 4
Stressing from 1 to 4
```
If we display the threads of  cloud-hypervisor when the problem happens, we see duplicated vcpus threads with the same id. For example, this is the initial state right after I booted the VM:

```bash
$ ps -T -p $(pidof cloud-hypervisor | cut -d" " -f2)                                                                           
    PID    SPID TTY          TIME CMD                                                                                                                                         
 139077  139077 pts/13   00:00:00 cloud-hyperviso                                                                                                                             
 139077  139078 pts/13   00:00:00 vmm                                                                                                                                         
 139077  139079 pts/13   00:00:00 http-server                                                                                                                                 
 139077  139080 pts/13   00:00:00 vmm_signal_hand                                                                                                                             
 139077  139099 pts/13   00:00:01 vcpu0                                                                                                                                       
 139077  139100 pts/13   00:00:01 vcpu1                                                                                                                                       
 139077  139101 pts/13   00:00:01 vcpu2                                                                                                                                       
 139077  139102 pts/13   00:00:01 vcpu3                                                                                                                                       
 139077  139104 pts/13   00:00:00 __console                                                                                                                                   
 139077  139105 pts/13   00:00:00 __rng                                                                                                                                       
 139077  139106 pts/13   00:00:00 _disk0_q0                                                                                                                                   
 139077  139108 pts/13   00:00:00 _disk1_q0                                                                                                                                   
 139077  139131 pts/13   00:00:00 iou-wrk-139106
 139077  139142 pts/13   00:00:00 iou-wrk-139108
 139077  139170 pts/13   00:00:00 iou-wrk-139106
 139077  139173 pts/13   00:00:00 iou-wrk-139106
 139077  139182 pts/13   00:00:00 iou-wrk-139106
```

Then, after stressing the VM until it becomes unresponsive, I observed the following:

```bash
$ ps -T -p $(pidof cloud-hypervisor | cut -d" " -f2)
    PID    SPID TTY          TIME CMD
 139077  139077 pts/13   00:00:00 cloud-hyperviso
 139077  139078 pts/13   00:00:00 vmm
 139077  139079 pts/13   00:00:00 http-server
 139077  139080 pts/13   00:00:00 vmm_signal_hand
 139077  139099 pts/13   00:00:01 vcpu0
 139077  139102 pts/13   00:00:01 vcpu3                  <==========
 139077  139104 pts/13   00:00:00 __console 
 139077  139105 pts/13   00:00:00 __rng
 139077  139106 pts/13   00:00:00 _disk0_q0 
 139077  139108 pts/13   00:00:00 _disk1_q0 
 139077  139131 pts/13   00:00:00 iou-wrk-139106
 139077  139142 pts/13   00:00:00 iou-wrk-139108
 139077  139189 pts/13   00:00:00 iou-wrk-139106
 139077  139207 pts/13   00:00:00 vcpu3                  <==========
 139077  139213 pts/13   00:00:00 vcpu3                  <==========
 139077  139219 pts/13   00:00:00 vcpu2
 139077  139220 pts/13   00:00:00 vcpu3                  <==========
```
In the above output, we can see 4 `vcpu3` threads.

**Analysis**

Here is what I've analyzed so far in `gdb` and logs:

Taking the above output as an example, when the VM becomes unresponsive and we get to the state where there are multiple `vpcu3` threads, `vpcu0` is stuck at `state.signal_thread()` loop https://github.com/cloud-hypervisor/cloud-hypervisor/blob/aa14fe214a8c02d980f742e5200323505fe03595/vmm/src/cpu.rs#L1127 because `vcpu_run_interrupted` never gets set to `true` for `vpcu3`. This occurs because the call to `libc::pthread_kill(handle.as_pthread_t() as _, SIGRTMIN())` https://github.com/cloud-hypervisor/cloud-hypervisor/blob/aa14fe214a8c02d980f742e5200323505fe03595/vmm/src/cpu.rs#L559 won't cause a VM exit for `vpcu3`, since the signal is sent to the wrong thread (PID 139220, the most recent created thread), not to the one actually hosting `vpcu3` (PID 139102).

In the above example (after the VM is hung), I was able to unhang the VM by sending some SIGRTMIN signals as follows:

1) First, we stop the script stressing the VM to avoid reaching the problem again.

2) Then we send SIGRTMIN to the oldest `vpcu3`, PID 139102 in the above output.
```bash
kill -SIGRTMIN 139102
```
This causes the `vpcu0` thread to break out from the `signal_thread()` loop, since `vcpu_run_interrupted` gets set to `true`.

3) However, then the execution of `vcpu0` stops at the line `state.join_thread()` https://github.com/cloud-hypervisor/cloud-hypervisor/blob/aa14fe214a8c02d980f742e5200323505fe03595/vmm/src/cpu.rs#L1128 because the current thread pointed by `vcpu_states[3].handle` is not the one we just interrupt (PID 139102). Thus, if we send SIGRTMIN to the other vpcus with id 3, from oldest to newest,  the VM will finally unhang:
```bash
kill -SIGRTMIN 139207
kill -SIGRTMIN 139213
kill -SIGRTMIN 139220
```
That happens because:
- Right after 139102 dies, 139207 is now hosting `vpcu3`. When we send `SIGRTMIN` to 139207, we have a VM exit, and since `vcpu_run_interrupted` is still `true` for `vpcu3` (I didn't find where it gets cleared), it will die.
- The same will now apply to 139213.
- Finally, when we send `SIGRTMIN` to 139220, the call to `handle.join()` returns, since now 139220 is hosting `vpcu3`, then we have a  VM exit with `vcpu_run_interrupted` set to `true`, which will make 139220 to end, thus allowing `vpcu0` to continue its execution after`state.join_thread()`.

**To Reproduce**
```bash
while true; do
  for i in $(seq 1 4); do
    curl  -X PUT -d"{\"desired_vcpus\": $i}" -H 'Accept: application/json' --unix-socket /tmp/cloud-hypervisor.sock  http://localhost/api/v1/vm.resize; 
  done
done
```

**Version**

Output of `cloud-hypervisor --version`:
v31.0-44-g97012c51-dirty

**VM configuration**

```json
{
        "balloon": null,
        "console": {
            "file": null,
            "iommu": false,
            "mode": "Tty"
        },
        "cpus": {
            "affinity": null,
            "boot_vcpus": 4,
            "features": {
                "amx": false
            },
            "kvm_hyperv": false,
            "max_phys_bits": 46,
            "max_vcpus": 4,
            "topology": null
        },
        "devices": null,
        "disks": [
            {
                "direct": false,
                "disable_io_uring": false,
                "id": "_disk0",
                "iommu": false,
                "num_queues": 1,
                "path": "/home/rsoares/repo/tmp/tests/new-image/focal-server-cloudimg-amd64.raw",
                "pci_segment": 0,
                "queue_size": 128,
                "rate_limiter_config": null,
                "readonly": false,
                "vhost_socket": null,
                "vhost_user": false
            },
            {
                "direct": false,
                "disable_io_uring": false,
                "id": "_disk1",
                "iommu": false,
                "num_queues": 1,
                "path": "/tmp/ubuntu-cloudinit.img",
                "pci_segment": 0,
                "queue_size": 128,
                "rate_limiter_config": null,
                "readonly": false,
                "vhost_socket": null,
                "vhost_user": false
            }
        ],
        "fs": null,
        "iommu": false,
        "memory": {
            "hotplug_method": "Acpi",
            "hotplug_size": null,
            "hotplugged_size": null,
            "hugepage_size": null,
            "hugepages": false,
            "mergeable": false,
            "prefault": false,
            "shared": false,
            "size": 1073741824,
            "thp": true,
            "zones": null
        },
        "net": null,
        "numa": null,
        "payload": {
            "cmdline": "console=hvc0 root=/dev/vda1 rw",
            "firmware": null,
            "initramfs": null,
            "kernel": "/home/rsoares/repo/linux-cloud-hypervisor/arch/x86/boot/compressed/vmlinux.bin"
        },
        "platform": null,
        "pmem": null,
        "rng": {
            "iommu": false,
            "src": "/dev/urandom"
        },
        "serial": {
            "file": null,
            "iommu": false,
            "mode": "Null"
        },
        "sgx_epc": null,
        "tpm": null,
        "user_devices": null,
        "vdpa": null,
        "vsock": null,
        "watchdog": false
}
```

Guest OS version details: 6.2.0+

Host OS version details: 5.18.2

### PR DESCRIPTION
**Pull request by up2wing**
Add pending removed vcpu check according to VcpuState.removing, which can avoid cloud hypervisor hangup during continual vcpu resize.

Fix #5419

### CODE REVIEW

File: vmm/src/cpu.rs [hunk 1]
```diff
@@ -518,12 +521,6 @@ impl BusDevice for CpuManager {
                     {
                         state.inserting = false;
                     }
-                    // Ditto for removal
-                    if (data[0] & (1 << CPU_REMOVING_FLAG) == 1 << CPU_REMOVING_FLAG)
```
Comment (rbradford) [inline vmm/src/cpu.rs:522]: Hhm, I think we might still want to keep the ACPI pending removal flag - my only worry is if the kernel gets confused that is has acknowledged the removal and then the removal bit doesn't get cleared when it checks again. Maybe we need to track another bit of state :-(

**Review by rbradford — COMMENTED** [review]
Thanks for diving in!

File: vmm/src/cpu.rs (hunk 1 above)
Comment (up2wing) [reply vmm/src/cpu.rs:522]: Okay. Let me think a better way out.

Comment (rbradford) [reply vmm/src/cpu.rs:522]: I think it's fine to keep the existing `removing` bool. And add a `Arc<AtomicBool>>` called `pending_removal`. Set both on the request to remove, let the guest clear the `removing` as it sees fit but use the atomic one to check if a subsequent removal should be rejected.

Comment (up2wing) [reply vmm/src/cpu.rs:522]: Thx for ur suggestions, that sounds make sense and I will update the patch.

**Review by rbradford — APPROVED** [review]

**Pull request merged**